packages/sdk-js/src/plugins/devtools.ts

```diff
@@ -0,0 +1,229 @@
+  ) {
+    gb.setForcedVariations(devtoolsState.experiments);
+  }
+}
+
+export function devtoolsPlugin(devtoolsState?: DevtoolsState): Plugin {
+  return (gb: GrowthBook | UserScopedGrowthBook | GrowthBookClient) => {
+    // Only works for standard GrowthBook instances
+    if ("createScopedInstance" in gb) {
+      return;
+    }
+    if (devtoolsState) {
+      applyDevtoolsState(devtoolsState, gb);
+    }
+  };
+}
+
+/**
+ * For NextJS environments.
+ * When using server components, use the `searchParams` and `requestCookies` fields.
+ *  - Note: In NextJS 15+, you should await these values before passing them to the plugin
+ * When using middleware / api routes, provide the `request` field instead.
+ */
+export function devtoolsNextjsPlugin({
+  searchParams,
+  requestCookies,
+  request,
+}: {
+  searchParams?: { _gbdebug?: string };
+  requestCookies?: NextjsReadonlyRequestCookiesCompat;
+  request?: NextjsRequestCompat;
+}): Plugin {
+  function extractGbDebugPayload({
+    searchParams,
+    requestCookies,
+  }: {
+    searchParams?: { _gbdebug?: string } | URLSearchParams;
+    requestCookies?: NextjsReadonlyRequestCookiesCompat;
+  }): string | undefined {
+    if (searchParams) {
+      if ("_gbdebug" in searchParams) {
+        return searchParams._gbdebug;
+      }
+      if (searchParams instanceof URLSearchParams) {
+        return searchParams.get("_gbdebug") ?? undefined;
+      }
+    }
+    return requestCookies?.get("_gbdebug")?.value;
+  }
+
+  return (gb: GrowthBook | UserScopedGrowthBook | GrowthBookClient) => {
+    let payload = extractGbDebugPayload({ searchParams, requestCookies });
+
+    if (!payload && request) {
+      payload = extractGbDebugPayload({
+        searchParams: request.nextUrl.searchParams,
+        requestCookies: request.cookies,
+      });
+    }
+
+    let state: DevtoolsState = {};
+    if (payload) {
+      try {
+        state = JSON.parse(payload);
+      } catch (e) {
+        console.error("cannot parse devtools payload", e);
+      }
+    }
+
+    devtoolsPlugin(state)(gb);
+  };
+}
+
+/**
+ * Intended to be used with cookieParser() middleware from npm: 'cookie-parser'.
+ */
+export function devtoolsExpressPlugin({
+  request,
+}: {
+  request?: ExpressRequestCompat;
+}): Plugin {
+  return (gb: GrowthBook | UserScopedGrowthBook | GrowthBookClient) => {
+    let payload =
+      typeof request?.query?.["_gbdebug"] === "string"
+        ? request.query["_gbdebug"]
+        : undefined;
+    if (!payload) {
+      payload =
+        typeof request?.cookies?.["_gbdebug"] === "string"
+          ? request.cookies["_gbdebug"]
+          : undefined;
+    }
+
+    let state: DevtoolsState = {};
+    if (payload) {
+      try {
+        state = JSON.parse(payload);
+      } catch (e) {
+        console.error("cannot parse devtools payload", e);
+      }
+    }
+
+    devtoolsPlugin(state)(gb);
+  };
+}
+
+type LogUnionWithSource = LogUnion & { source?: string };
```

Comment:
Ideally `source` would live within `SdkInfo` instead since it will be identical for all log entries.  I know the published devtools extension expects it this way though, so kind of annoying to change now.

Reply:
agree, only a minor change in devtools